stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -18,227 +18,346 @@ namespace stan {
 namespace math {
 namespace internal {
 
+/**
+ * Var specialization of implimentation called by reduce `reduce_sum`.
+ * @tparam ReduceFunction An type with a valid `operator()`
+ * @tparam ReturnType The return type of the call to
+ * `ReduceFunction::operator()`
+ * @tparam Vec type of the first input argument with an `operator[]`
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
 template <typename ReduceFunction, typename ReturnType, typename Vec,
           typename... Args>
 struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
                        Vec, Args...> {
-  template <typename T, require_arithmetic_t<scalar_type_t<T>>...>
-  static inline const T& deep_copy(const T& arg) {
-    return arg;
-  }
+  /**
+   * Internal object used in `tbb::parallel_reduce`
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
+   */
+  struct recursive_reducer {
+    size_t num_shared_terms_;  // Number of terms shared across threads
+    double* sliced_partials_;  // Points to adjoints of the partial calculations
+    Vec vmapped_;
```

Comment:
So if I pass in a ```const std::vector<T>&```, Vec will be a ```const std::vector<T>&```?

What if I pass in a ```std::vector<T>```? Will I get a full vector here or a reference?

Reply:
Since this is used in parallel reduce, and it's signature is all references, the types we can get back are 

```cpp
std::vector<T>&
const std::vector<T>&
std::vector<T>&&
```

If you passed in an `std::vector<T>` to reduce_sum_impl then `Var` would be equal to `std::vector<T>`